### ISSUE TITLE
Inconsistency in getting nodes by URI idType

### ISSUE THREAD

**Issue by TylerBarnes** [wp-graphql/wp-graphql#1209]
Some node types seem to be inaccessible by their URI. See the screenshot below.
User and TermNode work while Page and Post don't

<img width="732" alt="Screen Shot 2020-04-01 at 3 57 18 PM" src="https://user-images.githubusercontent.com/14190743/78194301-9cbbe300-7431-11ea-9d28-fcaf3cf2b185.png">

**Comment by TylerBarnes** [comment]
Oops, that was on my leading-slash uri branch. Same issue on master:

<img width="691" alt="Screen Shot 2020-04-01 at 4 00 59 PM" src="https://user-images.githubusercontent.com/14190743/78194452-09cf7880-7432-11ea-9c10-6bff858ab293.png">

**Comment by jasonbahl** [comment]
@TylerBarnes this is strange indeed. I'm not able to reproduce. 

Here are several attempts with various slashes. . .leading slash, trailing slash, neither, both. . .

![Screen Shot 2020-04-01 at 9 25 23 PM](https://user-images.githubusercontent.com/1260765/78207659-8f1d5200-745f-11ea-9f8c-23d3c36cc15f.png)
![Screen Shot 2020-04-01 at 9 25 34 PM](https://user-images.githubusercontent.com/1260765/78207662-904e7f00-745f-11ea-93af-7cc01cbb4602.png)
![Screen Shot 2020-04-01 at 9 25 46 PM](https://user-images.githubusercontent.com/1260765/78207663-90e71580-745f-11ea-98fb-9803383acf45.png)
![Screen Shot 2020-04-01 at 9 25 59 PM](https://user-images.githubusercontent.com/1260765/78207665-917fac00-745f-11ea-8099-a5608a481424.png)

**Comment by jasonbahl** [comment]
@TylerBarnes can you maybe try flushing your permalinks and trying again? If that doesn't help, maybe we can pair on this and see what we can find out.

**Comment by jasonbahl** [comment]
@TylerBarnes and I confirmed that this seems related to permalink structures. This needs more testing with various permalink settings.

**Comment by jonshipman** [comment]
I noticed that CPT UI, the way it flushes permalinks, leaves them at the end of $wp_rewrite->wp_rewrite_rules(). So the ```// Got a match``` in the src/Data/NodeResolver.php breaks after [^/]+/([^/]+)/?$ instead of location\/([^\/]+)(?:\/([0-9]+))?\/?$ (when dealing with a custom post type with the post_type "location."

If you remove the break on line 153 of NodeResolver you get the proper match in $this->wp->matched_rule but it does not fix the issue. $this->wp->query_vars on line 219 remains empty and $this->wp->query_vars['name'] is an empty string after the loop.

In an unmodified NodeResolver, for the uri /location/kansas-city I see "attachment=kansas-city" for $perma_query_vars. Removing the break I see "name=&feed=" for $perma_query_vars.

**Comment by jonshipman** [comment]
I can at least confirm in my case that the standard WP::parse_request has the same issue. Either a bug in WP Core, CPT UI, WP-GraphQL, or all of the above.

**Comment by jonshipman** [comment]
```
add_filter(
    'cptui_pre_register_post_type',
    function( $args, $post_type_name ) {
        if ( 'app_page' === $post_type_name ) {
            $args['rewrite']['slug'] = '/';
        }

        return $args;
    },
    10,
    2
);
```

I found if you have a custom post-type and rewrite the slug to be a page level item, it will be processed as an empty string. This seems to be a WP Core related item (understandable as resolve_uri is mostly WP::parse_request).

Perhaps there might be some benefit to WPGraph-QL in somehow using the built-in WP::parse_request as opposed to duplicating it?

**Comment by TylerBarnes** [comment]
@jonshipman this is interesting, awesome work here. Do you know of an easy way to reproduce this bug using CPTUI? I'm having trouble reproducing it now 🤔

**Comment by jonshipman** [comment]
@TylerBarnes 

Yeah, add the filter from my last comment and change "app_page" to the post type of one of your CPTUI post types.

**Comment by jasonbahl** [comment]
@jonshipman has this been resolved with the work we did on URIs recently when I paired with you?

**Comment by jasonbahl** [comment]
@jonshipman 
> Perhaps there might be some benefit to WPGraph-QL in somehow using the built-in WP::parse_request as opposed to duplicating it?

I tried that initially, but if I recall correctly, it's not intended to be re-used and there's no real good way to re-use it as-is.

**Comment by jonshipman** [comment]
I believe my specific issue was actually a core WordPress issue. I tried Tyler's query with posts and had no issue.

![image](https://user-images.githubusercontent.com/12699954/98412992-823c8e80-203e-11eb-82dc-df7459485d65.png)

The issue with the slide in that example is because I changed the rewrite slug to '/'. Doesn't seem like WordPress supports that and they don't have an error on their end concerning it. Seems like an edge case TBH.
